[RFC] Fix relative paths on Windows (#7117)

aignas opened the pull request:
See #7117.

I am creating the PR in order to benefit from the CI setup as this is my first PR.

aignas committed:
Add changes from a comment by justinmk (#7117)

Refer to
https://github.com/neovim/neovim/issues/2471#issuecomment-271193714

aignas committed:
Add tests to reproduce #7117

aignas commented:
The test failures:
```
[----------] Running tests from C:/projects/neovim/test/functional\core\path_spec.lua
[ RUN      ] 'path' given a relative path with current directory to a file expanding should work correctly #7117: FAIL
C:/projects/neovim/test/functional\core\path_spec.lua:27: Expected objects to be the same.
Passed in:
(string) 'C:\projects\neovim\test\functional\fixtures\.\tty-test.c'
Expected:
(string) 'C:\projects\neovim\test\functional\fixtures\tty-test.c'
stack traceback:
	C:/projects/neovim/test/functional\core\path_spec.lua:27: in function <C:/projects/neovim/test/functional\core\path_spec.lua:23>
[ RUN      ] 'path' given a relative path with directory up the tree to a file expanding should work correctly #7117: FAIL
C:/projects/neovim/test/functional\core\path_spec.lua:36: Expected objects to be the same.
Passed in:
(string) 'C:\projects\neovim\test\functional\fixtures\..\fixtures\tty-test.c'
Expected:
(string) 'C:\projects\neovim\test\functional\fixtures\tty-test.c'
stack traceback:
	C:/projects/neovim/test/functional\core\path_spec.lua:36: in function <C:/projects/neovim/test/functional\core\path_spec.lua:32>
[----------] 2 tests from C:/projects/neovim/test/functional\core\path_spec.lua (94.00 ms total)
[----------] Running tests from C:/projects/neovim/test/functional\eval\api_functions_spec.lua
[ RUN      ] api functions work: 0.00 ms OK
```

It is interesting enough, that when the autochdir option is on (see 4b045aa), the path via `getcwd` was resolved correctly.

aignas commented:
The failures now:
```
[----------] Running tests from C:/projects/neovim/test/functional\core\path_spec.lua
[ RUN      ] '%:p' expanding given a relative path with current directory in the middle #7117: FAIL
C:/projects/neovim/test/functional\core\path_spec.lua:31: Expected objects to be the same.
Passed in:
(string) 'C:\projects\neovim\test\functional\fixtures\.\tty-test.c'
Expected:
(string) 'C:\projects\neovim\test\functional\fixtures\tty-test.c'
stack traceback:
	C:/projects/neovim/test/functional\core\path_spec.lua:31: in function <C:/projects/neovim/test/functional\core\path_spec.lua:29>
[ RUN      ] '%:p' expanding given a relative path with current directory #7117: FAIL
C:/projects/neovim/test/functional\core\path_spec.lua:36: Expected objects to be the same.
Passed in:
(string) 'C:\projects\neovim\.\test\functional\fixtures\tty-test.c'
Expected:
(string) 'C:\projects\neovim\test\functional\fixtures\tty-test.c'
stack traceback:
	C:/projects/neovim/test/functional\core\path_spec.lua:36: in function <C:/projects/neovim/test/functional\core\path_spec.lua:34>
[ RUN      ] '%:p' expanding given a relative path with current directory to a file when changing directory #7117: FAIL
C:/projects/neovim/test/functional\core\path_spec.lua:42: Expected objects to be the same.
Passed in:
(string) 'C:\projects\neovim\test\C:\projects\neovim\.\test\functional\fixtures\tty-test.c'
Expected:
(string) 'C:\projects\neovim\test\functional\fixtures\tty-test.c'
stack traceback:
	C:/projects/neovim/test/functional\core\path_spec.lua:42: in function <C:/projects/neovim/test/functional\core\path_spec.lua:39>
[ RUN      ] '%:p' expanding given a relative path with directory up the tree to a file #7117: FAIL
C:/projects/neovim/test/functional\core\path_spec.lua:47: Expected objects to be the same.
Passed in:
(string) 'C:\projects\neovim\test\functional\fixtures\..\fixtures\tty-test.c'
Expected:
(string) 'C:\projects\neovim\test\functional\fixtures\tty-test.c'
stack traceback:
	C:/projects/neovim/test/functional\core\path_spec.lua:47: in function <C:/projects/neovim/test/functional\core\path_spec.lua:45>
[ RUN      ] '%:p' expanding given a different starting directory and a relative path with directory up the tree #7117: 0.00 ms OK
[ RUN      ] '%:p' expanding given a different starting directory and a relative path with current directory and up the tree #7117: 0.00 ms OK
[----------] 6 tests from C:/projects/neovim/test/functional\core\path_spec.lua (265.00 ms total)
```

justinmk commented:
The tests pass on non-windows, this suggests there's some logic somewhere remove the `/./` and `/../`. But we need to change that logic to be path-separator agnostic.

justinmk commented:
`path_get_absolute_path` looks like one obvious culprit. The `/` literals should be `PATHSEP`.

aignas commented:
@justinmk, yeah, after seeing the test results I thought of the same. Will looke into these and thanks for letting me know of the PATHSEP macro. :)

aignas committed:
Start replacing / with PATHSEP

aignas commented:
@justinmk, the `path_spec.lua` tests have passed and the tests that failed on Windows seem to be unrelated to my changes, but it would be good if you could have a look. Let me know if you want any additional code changes to be made.

justinmk commented:
Thanks. This test needs a fixup:

```
[  ERROR   ] C:/projects/neovim/test/functional\ui\screen_basic_spec.lua @ 121: Screen :set title has correct default title with named file
.\test\functional\ui\screen.lua:302: C:/projects/neovim/test/functional\ui\screen_basic_spec.lua:126: Expected objects to be the same.
Passed in:
(string) 'myfile (C:\projects\neovim\\mydir) - NVIM'
Expected:
(string) 'myfile (/mydir) - NVIM'
stack traceback:
```

The shada test failures will need to be fixed as well. For example:

```

[  FAILED  ] C:/projects/neovim/test/functional\shada\merging_spec.lua @ 815: ShaDa jumps support code merges jumps when reading
C:/projects/neovim/test/functional\shada\merging_spec.lua:825: Expected objects to be the same.
Passed in:
(string) '
 jump line  col file/text
   1     1    0 
>'
Expected:
(string) '
 jump line  col file/text
   6     2    0 /a/b/c
   5     2    0 /a/b/d
   4     3    0 /a/b/d
   3     2    0 /a/b/e
   2     2    0 /a/b/f
   1     1    0 
>'
stack traceback:
	C:/projects/neovim/test/functional\shada\merging_spec.lua:825: in function <C:/projects/neovim/test/functional\shada\merging_spec.lua:815>
```

In that test, `wshada()` takes a path like `.../a/b/c/...`. That's stored in the shada file literally. Most likely somewhere in the C code the path isn't getting translated correctly. Though after looking, I cannot see where.

aignas committed:
Fixup tests: ui/screen_basic

aignas committed:
Fixup tests: shada/merging and shada/compatibility

Currently using backward slashes in the tests does not work.
It also seems that the maximum length of the file path is only 6
characters as there may be a bug in the shada parsing test code.

justinmk commented on test/functional/shada/merging_spec.lua:
```diff
@@ -13,6 +13,11 @@ local read_shada_file = shada_helpers.read_shada_file
+local mock_file_path = '/a/b/'
+if helpers.iswin() then
+  mock_file_path = 'C:\\a\\b\\'
+end
```
I'm not sure this is the right approach. If a path with `/` slashes gets stored in shada, nvim should still be able to understand it. @ZyX-I what do you think?

aignas replied:
I agree. However, I was trying to construct a payload with a filename, that has more than 6 characters in the path and as far as I understand the payload part with the filename should be something like `\xc4\x07/a/bb/c` but that does not work, however `\xc4\x05/aa/c` does work, which leaves me completely baffled. I was reading on the [bin format in msgpack spec](https://github.com/msgpack/msgpack/blob/master/spec.md#bin-format-family).

The example test file:
```lua
-- ShaDa filename length greater that 6 test
local helpers = require('test.functional.helpers')(after_each)
local exc_exec = helpers.exc_exec
local eq = helpers.eq
local shada_helpers = require('test.functional.shada.helpers')
local reset, clear, get_shada_rw =
  shada_helpers.reset, shada_helpers.clear, shada_helpers.get_shada_rw
local wshada, sdrcmd, shada_fname =
  get_shada_rw('Xtest-functional-shada-merging.shada')

describe('When trying to par ShaDa code this', function()
  before_each(reset)
  after_each(function()
    clear()
    os.remove(shada_fname)
  end)

  it('works #7287',
  function()
    wshada('\007\001\018\131\162mX\195\161f\xc4\006/a/b/-\161nA')
    eq(0, exc_exec(sdrcmd()))
  end)

  it('does not work #7287',
  function()
    wshada('\007\001\018\131\162mX\195\161f\xc4\007/a/bc/-\161nA')
    eq(0, exc_exec(sdrcmd()))
  end)
end)
```

Do you know if there is any documentation on the protocol that NVIM uses in order to store things?

ZyX-I commented:
Full documentation on ShaDa is in `:h shada-format`, unlike viminfo this is considered to be a part of API. It does not contain documentation on msgpack though.

aignas commented:
I think that there may be a bug with the test code as I can successfully unpack the following byte-string using python's umsgpack:
```
In [5]: umsgpack.unpackb(b'\x83\xa2mX\xc3\xa1f\xc4\x08/a/bbc/-\xa1nA')
Out[5]: {'f': b'/a/bbc/-', 'mX': True, 'n': 65}
```

This is also decoded correctly when reading the shada file in nvim. However, when I am trying to read the shada using the `exc_exec` helper from lua, it breaks. I noticed that in all of the tests the file path is 6 characters or less.

Can anyone confirm that it breaks if you have 7 characters or more? That may be a problem because an equivalent test for windows might need a path, that is 8 characters long (I could make it 6 as a work around to test my theory.

aignas commented:
In the latest appveyor build we have the tests passing, but one build failed at the compilation stage. Hence, I have rebased on `origin/master` and squashed the test fixup commits.

justinmk commented:
@aignas Nice, I'll look at this soon. If you have a 64bit windows system to test on, you could download the artifact from https://ci.appveyor.com/project/neovim/neovim/build/15371 (click the MINGW_64 build, then "Artifacts" tab) and see if it works as expected.

aignas commented:
@justinmk, just tested on a Windows 64-bit machine and it seems to behave as expected. :) At least the bug in #7117 seems to be working.

justinmk commented:
> we need to revisit do_path_expand and compare Vim's dos_expandpath() and unix_expandpath().

Done, found a couple minor differences which I will put in another PR.

justinmk commented on src/nvim/path.c:
```diff
@@ -2196,11 +2202,11 @@ static int path_get_absolute_path(const char_u *fname, char_u *buf,
   // expand it if forced or not an absolute path
   if (force || !path_is_absolute_path(fname)) {
-    if ((p = vim_strrchr(fname, '/')) != NULL) {
+    if ((p = vim_strrchr(fname, PATHSEP)) != NULL) {
```
It looks like `slash_adjust()` isn't always guaranteed to have already been called by now, so there could be a `/` slash here on Windows. It  probably makes sense to check for both here. Maybe we need a new function in the style of `slash_adjust` :

(untested...)

```c
char *slash_index(char *fname)
{
  char *p = NULL;
#ifdef BACKSLASH_IN_FILENAME
  p = vim_strrchr(psepc);
  if (p == NULL) {
    p = vim_strrchr(psepcN);
  }
#else
  p = vim_strrchr(PATHSEP);
#endif
  return p;
}
```

`path_fix_case` has another `vim_strrchr(name, '/')` case.

In these cases

justinmk replied:
Not sure about this though , so maybe different PR.

aignas replied:
When would that would be the case? Because if we can write a test for it and this is inside the scope of the expand('%:p') case, then I would be willing to do it before the merge if it is a different case, then it is maybe better to open an issue and address it into a separate PR. :)

justinmk replied:
I don't know. Finding a test case for this is the hard part :)

janlazo commented:
Is `glob` within the scope of this PR for relative paths?

aignas commented:
I would say that it would be nice to make `glob` separately, but since we are at it, we could do it as a part of this PR, since the name of the PR is quite generic "Fix relative paths on Windows", but let's define the scope of this now. :)

@janlazo, do you have any examples of the incorrect behaviour for glob? The current test coverage at `test/functional/eval` for `glob` is not great, so it would be nice to first write some tests. I see that there are some tests that are disabled on windows in `/home/ia/src/github/neovim/test/functional/legacy/097_glob_path_spec.lua`, so I can enable them and see what is breaking.

@justinmk, as I understand the outstanding issues are:
- [ ] Change `path_get_absolute_path` to work on Windows if `slash_adjust` has not been called yet.
- [ ] Review `do_expand_path` and `path_fix_case` and make them consistent on Windows.

Am I missing anything?

janlazo commented:
@aignas I don't have examples of incorrect behaviour of glob. A test case I want to add is `glob(tempname())` to test for 8.3 short filename in DOS but `expand('%:p:8')` doesn't work yet. Workaround in https://github.com/junegunn/fzf/issues/1018#issuecomment-330391753 to generate it via cmd.exe

justinmk commented:
> A test case I want to add is glob(tempname()) to test for 8.3 short filename in DOS but expand('%:p:8') doesn't work yet. 

@janlazo Nvim removed support for DOS short names.

@aignas I think https://github.com/neovim/neovim/pull/7287#issuecomment-331936765 is correct, the remaining items can wait for another PR(s).

janlazo commented:
It is removed in https://github.com/neovim/neovim/pull/635 😢 . It is a convenient workaround when double-quoting is insufficient for escaping paths so it's used in fzf.vim for its preview script.

Will `tempname` eventually use the full unmodified path instead of DOS shortname?

justinmk commented:
> It is a convenient workaround when double-quoting is insufficient for escaping paths so it's used in fzf.vim for its preview script.

Oh wow. Spacebar heating ... :)

> Will tempname eventually use the full unmodified path instead of DOS shortname?

Didn't know about that, we should definitely change that.

janlazo commented:
If a file named `$TMP` is directly on the working directory, what should the output of `expand('$TMP')` and `glob('$TMP')` be?

justinmk commented:
> If a file named $TMP is directly on the working directory, what should the output of expand('$TMP') and glob('$TMP') be?

Do we do something different than Vim?

ZyX-I commented on test/functional/shada/merging_spec.lua:
```diff
@@ -13,6 +13,12 @@ local read_shada_file = shada_helpers.read_shada_file
   get_shada_rw('Xtest-functional-shada-merging.shada')
 
+local mock_file_path = '/a/b/'
+local header = '\xc4\x06'
```
AFAIR `\xXX` is supposed to be unavailable in some circumstances. And in any case this is inconsistent with the rest of the file.

justinmk replied:
Thanks @ZyX-I . @aignas FYI I reverted this in the merge.

ZyX-I commented on test/functional/shada/merging_spec.lua:
```diff
@@ -569,15 +575,15 @@ describe('ShaDa marks support code', function()
     local found = 0
     for _, v in ipairs(read_shada_file(shada_fname)) do
-      if v.type == 7 and v.value.f == '/a/b/-' then
+      if v.type == 7 and v.value.f == '' .. mock_file_path .. '-' then
```
`:%s/'' \.\. //g`

ZyX-I commented on test/functional/shada/merging_spec.lua:
```diff
@@ -13,6 +13,12 @@ local read_shada_file = shada_helpers.read_shada_file
   get_shada_rw('Xtest-functional-shada-merging.shada')
 
+local mock_file_path = '/a/b/'
+local header = '\xc4\x06'
```
I do not see where `header` variable changes, so what’s the point in its existence at all?

justinmk commented on test/functional/core/path_spec.lua:
```diff
@@ -0,0 +1,61 @@
+    clear()
+    pathsep = get_pathsep()
+    targetdir = join_path('test', 'functional', 'fixtures')
+    clear(join_path(targetdir, 'tty-test.c'))
```
`clear()` is somewhat expensive, specially to do it +2 times per test. So I removed the first case and move the other case to `setup()` (which only runs once per `describe()` block)

justinmk commented:
Merged in https://github.com/neovim/neovim/pull/7349 . @aignas thanks for your attention to this!
